src/peft/tuners/xlora/classifier.py

```diff
@@ -0,0 +1,257 @@
+
+                result: ModelOutput = model.forward(
+                    *args,
+                    input_ids=input_ids,
+                    inputs_embeds=inputs_embeds,
+                    _xlora_classifier_inhibitor_flag=InhibitorFlagPayload(
+                        batch_size=batch_size,
+                        seq_len=seq_len,
+                        override_scaling_pass_value=self.override_scaling_pass_value,
+                    ),
+                    **kwargs,
+                )
+
+        hidden_states = result.hidden_states  # type: ignore
+
+        assert hidden_states is not None
+        hidden_state = hidden_states[-1]  # Get the last hidden state
+
+        ### Classifier run
+        # hidden_state=[batch_size, seq_len, hidden_size]
+        for layer in self.inner:
+            hidden_state = layer.forward(hidden_state)
+
+        logits = self.last.forward(hidden_state)
+
+        ### Repeat to make layerwise scalings if the classifier layer does not
+        if not self.config.layerwise_scalings:
+            logits = logits.unsqueeze(2)
+            logits = logits.expand(-1, -1, self.n_layers, -1)
+
+        ### Classifier run
+
+        scalings = logits.reshape(batch_size, seq_len, self.n_layers, self.n_classes)
+        # scalings = [batch_size, seq_len, n_layers, n_classes]
+
+        if self.config.enable_softmax:
+            scalings = self.softmax(scalings)
+
+        if self.n_predictions_lifetime > 0:
+            print(f"Scaling predictions: {scalings}")
+            self.n_predictions_lifetime -= 1
+
+        if self.scalings_logging:
+            self.log_scalings.append(scalings)
+
+        return scalings
+
+    def get_nb_trainable_parameters(self):
+        # https://github.com/huggingface/peft/blob/main/src/peft/mixed_model.py#L156
+        r"""
+        Returns the number of trainable parameters and number of all parameters in the model.
+        """
+        trainable_params = 0
+        all_param = 0
+        for _, param in self.named_parameters():
+            num_params = param.numel()
+            # if using DS Zero 3 and the weights are initialized empty
+            if num_params == 0 and hasattr(param, "ds_numel"):
+                num_params = param.ds_numel  # type: ignore
+
+            # Due to the design of 4bit linear layers from bitsandbytes
+            # one needs to multiply the number of parameters by 2 to get
+            # the correct number of parameters
+            if param.__class__.__name__ == "Params4bit":
+                num_params = num_params * 2
+
+            all_param += num_params
+            if param.requires_grad:
+                trainable_params += num_params
+
+        return trainable_params, all_param
+
+    @staticmethod
+    def _save_scalings(file: str, scalings: List[torch.Tensor]):
+        result = torch.cat(scalings, dim=0)
+        npy = result.numpy()
+        numpy.save(file, npy)
+
+    def flush_log_scalings(self, path: str):
```

Comment:
Could you explain why we need this? Is this just for debugging/monitoring purposes?

Reply:
This API enables the user to get a log of the scalings. It is useful for generating visualizations such as [this](https://github.com/EricLBuehler/xlora/blob/master/res/token_by_token_scalings.gif).

Reply:
I see. In that case, there should be some kind of example provided to illustrate how to use this, I think otherwise users will not discover this feature.

Also, I'd suggest for this method to only return the `indices_map` and allow the caller to decide themselves if they want to save it on disk as json or do something else with it.

Reply:
Wouldn't returning the `indices_map` (or maybe the `seqlens_map` would be better, as it contains the tensors) make `flush_log_scalings` redundant to `get_scalings_log`? Of course, a dedicated method to calculate the `indices_map` may be helpful, and given the addition of such a method I think removing `flush_log_scalings` would be OK.

Additionally, there is actually a wrapper method on `XLoraModel` with the rest of the API which contains a docstring. This way, it should be easy to find. This method in the classifier is actually internal, I have prefixed it with `_` now.